File: archive/honey-crypto/src/threshold/pke.cc
#include "crypto/threshold/pke.hpp"
#include "blst/ops.hpp"
#include "crypto/error.hpp"
#include "crypto/threshold/blst/Scalar.hpp"
#include "threshold/interpolate.hpp"
#include "threshold/key_gen.hpp"
#include "utils.hpp"
#include <array>
#include <cstring>
#include <expected>
#include <openssl/rand.h>
#include <span>
#include <system_error>
#include <vector>

namespace Honey::Crypto::Threshold::Pke {

namespace {
    std::array<Byte, 32> xor_bytes(std::span<const Byte, 32> a, std::span<const Byte, 32> b)
    {
        std::array<Byte, 32> res;
        for (size_t i = 0; i < a.size(); ++i) {
            res[i] = a[i] ^ b[i];
        }
        return res;
    }
    /// 验证密钥封装的一致性
    ///
    /// e(w, G1) == e(H(u,v), u)
    bool verify_encapsulation(const Ciphertext& ciphertext)
    {
        // 映射参数: (q1, p1, q2, p2) -> (G2, G1, G2, G1)
        return bls::ops::verify_pairing_equality(
            ciphertext.w, // q1: w
            G1Point::generator(), // p1: G1_gen
            Utils::hashH(ciphertext.u, ciphertext.v), // q2: H(u,v)
            ciphertext.u // p2: u
        );
    }
} // namespace

/// 验证解密份额
///  e(G2, ui) == e(yi, u)
bool verify_share(const PublicParameters& public_params,
    const PartialDecryptionShare& decryption,
    const Ciphertext& ciphertext)
{
    auto id = decryption.player_id;
    if (id < 1 || id > public_params.total_players)
        return false;

    return bls::ops::verify_pairing_equality(
        G2Point::generator(), // q1: G2_gen
        decryption.value, // p1: ui (份额)
        public_params.verification_vector[id - 1], // q2: yi (公钥份额)
        ciphertext.u // p2: u (密钥封装中的u)
    );
}

auto generate_keys(uint32_t num_players, uint32_t threshold)
    -> std::expected<KeySet, std::error_code>
{
    return Threshold::generate_keys<MasterPublicKey, VerificationKey>(num_players, threshold);
}

Ciphertext seal(
    const MasterPublicKey& master_public_key,
    PlainText msg)
{
    auto random_scalar = *Scalar::random();

    G1Point u = G1Point::generator();
    bls::ops::mult(u, random_scalar);

    G1Point mask_point = master_public_key;
    bls::ops::mult(mask_point, random_scalar);
    auto mask = Utils::hashG(mask_point);

    auto v = xor_bytes(msg, mask);

    G2Point h = Utils::hashH(u, v);
    G2Point w = h;
    bls::ops::mult(w, random_scalar);

    Ciphertext result { .u = u, .v = v, .w = w };
    return result;
}

auto partial_open(
    const PrivateKeyShare& private_key_share,
    const Ciphertext& ciphertext) -> std::expected<PartialDecryptionShare, std::error_code>
{
    if (!verify_encapsulation(ciphertext)) {
        return std::unexpected(make_error_code(Crypto::Error::InvalidCiphertext));
    }
    // 生成解密份额
    auto share_value = ciphertext.u;
    bls::ops::mult(share_value, private_key_share.secret);

    return PartialDecryptionShare {
        .player_id = static_cast<PlayerIndex>(private_key_share.player_id),
        .value = share_value
    };
}

[[nodiscard]]
auto open(const PublicParameters& public_params,
    const Ciphertext& ciphertext,
    std::span<const PartialDecryptionShare> shares)
    -> std::expected<std::array<Byte, 32>, std::error_code>
{
    if (shares.size() < static_cast<size_t>(public_params.threshold)) {
        return std::unexpected(std::make_error_code(std::errc::message_size));
    }
    for (const auto& share : shares) {
        if (!verify_share(public_params, share, ciphertext)) {
            return std::unexpected(make_error_code(std::errc::bad_message));
        }
    }

    auto interpolation_result = Crypto::Math::interpolate_at_zero(shares);
    if (!interpolation_result) {
        return std::unexpected(interpolation_result.error());
    }
    const auto& recovered_point = *interpolation_result;

    // 恢复会话密钥: K = v ⊕ H_G(recovered_point)
    auto mask = Utils::hashG(recovered_point);

    std::array<Byte, 32> session_key;
    for (size_t i = 0; i < 32; ++i) {
        session_key[i] = ciphertext.v[i] ^ mask[i];
    }

    return session_key;
}

} // namespace Honey::Crypto::Threshold::Pke

/*                     \
 ===================== \
 Unit Test             \
 ==================    \
*/                     \
#ifndef DOCTEST_CONFIG_DISABLE
#include <array>
#include <doctest/doctest.h>
#include <string>
#include <vector>

namespace Honey::Crypto::Threshold::Pke {

namespace {

    std::array<Byte, 32> string_to_block32(const std::string& s)
    {
        std::array<Byte, 32> out {};
        std::memcpy(out.data(), s.data(), std::min<size_t>(32, s.size()));
        return out;
    }

} // namespace

struct TpkeTest {
    static constexpr uint32_t N = 5;
    static constexpr uint32_t K = 3;

    KeySet key_set;

    TpkeTest()
    {
        auto ks = generate_keys(N, K);
        REQUIRE(ks.has_value());
        key_set = *ks;
    }
};

TEST_CASE_FIXTURE(TpkeTest, "TPKE.seal_and_open_success")
{
    auto msg = string_to_block32("HoneyBadger BFT is robust!");

    Ciphertext ct = seal(key_set.public_params.master_public_key, msg);

    std::vector<PartialDecryptionShare> shares;
    for (int id : { 1, 3, 5 }) {
        auto s = partial_open(key_set.private_shares[id - 1], ct);
        REQUIRE(s.has_value());
        shares.push_back(*s);
    }

    auto opened = open(key_set.public_params, ct, shares);
    REQUIRE(opened.has_value());
    CHECK(*opened == msg);
}

TEST_CASE_FIXTURE(TpkeTest, "TPKE.open_fails_with_invalid_share")
{
    auto msg = string_to_block32("This should not decrypt");
    Ciphertext ct = seal(key_set.public_params.master_public_key, msg);

    std::vector<PartialDecryptionShare> shares;

    auto s1 = partial_open(key_set.private_shares[0], ct);
    auto s2 = partial_open(key_set.private_shares[1], ct);
    REQUIRE(s1.has_value());
    REQUIRE(s2.has_value());

    shares.push_back(*s1);
    shares.push_back(*s2);

    // 构造一个伪造份额
    PartialDecryptionShare bad {
        .player_id = 3,
        .value = G1Point::generator()
    };
    shares.push_back(bad);

    auto opened = open(key_set.public_params, ct, shares);
    CHECK_FALSE(opened.has_value());
}

TEST_CASE_FIXTURE(TpkeTest, "TPKE.open_fails_with_insufficient_shares")
{
    auto msg = string_to_block32("Not enough shares");
    Ciphertext ct = seal(key_set.public_params.master_public_key, msg);

    std::vector<PartialDecryptionShare> shares;
    auto s1 = partial_open(key_set.private_shares[0], ct);
    auto s2 = partial_open(key_set.private_shares[1], ct);
    REQUIRE(s1.has_value());
    REQUIRE(s2.has_value());

    shares.push_back(*s1);
    shares.push_back(*s2);

    auto opened = open(key_set.public_params, ct, shares);
    CHECK_FALSE(opened.has_value());
}

TEST_CASE_FIXTURE(TpkeTest, "TPKE.open_fails_with_duplicate_player_id")
{
    auto msg = string_to_block32("Duplicate share");
    Ciphertext ct = seal(key_set.public_params.master_public_key, msg);

    auto s = partial_open(key_set.private_shares[0], ct);
    REQUIRE(s.has_value());

    std::vector<PartialDecryptionShare> shares {
        *s,
        *s, // duplicate
        *s
    };

    auto opened = open(key_set.public_params, ct, shares);
    CHECK_FALSE(opened.has_value());
}
} // namespace Honey::Crypto::Threshold::Pke
#endif // DOCTEST_CONFIG_DISABLE

File: native/honey-native/src/crypto/threshold/pke.rs
use crate::crypto::crypto_error::CryptoError;

use super::{
    keygen::{Ciphertext, PartialDecryptionShare, PkePrivateKeyShare, PkePublicParams},
    utils::{hash_g, hash_h},
};
use crate::crypto::bls::{
    fr::Fr, g1::G1, g2::G2, interpolate::interpolate_at_zero, pairing::verify_pairing_equality,
};

/// Encrypt a 32-byte message under the master public key.
pub fn seal(mpk: &G1, msg: [u8; 32]) -> Ciphertext {
    let r = Fr::random();

    // U = r·G1
    let u = G1::generator().scalar_mult(&r);

    // mask = hashG(r·mpk)
    let mask_point = mpk.clone().scalar_mult(&r);
    let mask = hash_g(&mask_point);

    // V = msg ⊕ mask
    let mut v = [0u8; 32];
    for i in 0..32 {
        v[i] = msg[i] ^ mask[i];
    }

    // H = hashH(U, V), W = r·H
    let h = hash_h(&u, &v);
    let w = h.scalar_mult(&r);

    Ciphertext { u, v, w }
}

/// Verify that the ciphertext is well-formed: e(W, G1_gen) == e(H(U,V), U).
fn verify_encapsulation(ct: &Ciphertext) -> bool {
    let h = hash_h(&ct.u, &ct.v);
    // verify_pairing_equality(q1:G2, p1:G1, q2:G2, p2:G1)
    // e(w, G1_gen) == e(H(u,v), u)
    verify_pairing_equality(&ct.w, &G1::generator(), &h, &ct.u)
}

/// Verify a partial decryption share against the player's verification key.
pub fn verify_share(
    params: &PkePublicParams,
    share: &PartialDecryptionShare,
    ct: &Ciphertext,
) -> bool {
    let id = share.player_id;
    if id < 1 || id > params.total_players {
        return false;
    }
    let ver_key = &params.verification_vector[id - 1];
    // e(G2_gen, share_val) == e(ver_key, U)
    verify_pairing_equality(&G2::generator(), &share.value, ver_key, &ct.u)
}

/// Compute a partial decryption share for player i.
pub fn partial_open(
    share: &PkePrivateKeyShare,
    ct: &Ciphertext,
) -> Result<PartialDecryptionShare, CryptoError> {
    if !verify_encapsulation(ct) {
        return Err(CryptoError::InvalidCiphertext);
    }
    let value = ct.u.clone().scalar_mult(&share.secret);
    Ok(PartialDecryptionShare {
        player_id: share.player_id,
        value,
    })
}

/// Public API to verify ciphertext correctness.
pub fn verify_ciphertext(_params: &PkePublicParams, ct: &Ciphertext) -> Result<(), CryptoError> {
    if !verify_encapsulation(ct) {
        return Err(CryptoError::InvalidCiphertext);
    }
    Ok(())
}

/// Combine partial decryption shares to recover the plaintext.
pub fn open(
    params: &PkePublicParams,
    ct: &Ciphertext,
    shares: &[PartialDecryptionShare],
) -> Result<[u8; 32], CryptoError> {
    if shares.len() < params.threshold {
        return Err(CryptoError::InsufficientShares {
            need: params.threshold,
            got: shares.len(),
        });
    }

    for s in shares {
        if !verify_share(params, s, ct) {
            return Err(CryptoError::VerificationFailed);
        }
    }

    let pairs: Vec<(usize, G1)> = shares.iter().map(|s| (s.player_id, s.value)).collect();
    let recovered = interpolate_at_zero(&pairs)?;

    let mask = hash_g(&recovered);
    let mut msg = [0u8; 32];
    for i in 0..32 {
        msg[i] = ct.v[i] ^ mask[i];
    }
    Ok(msg)
}

#[cfg(test)]
mod tests {
    use super::super::keygen::{PkeKeySet, generate_pke_keys};
    use super::*;

    fn make_msg(s: &str) -> [u8; 32] {
        let mut out = [0u8; 32];
        let bytes = s.as_bytes();
        let len = bytes.len().min(32);
        out[..len].copy_from_slice(&bytes[..len]);
        out
    }

    struct Fixture {
        ks: PkeKeySet,
    }

    impl Fixture {
        fn new(n: usize, k: usize) -> Self {
            let ks = generate_pke_keys(n, k).unwrap();
            Fixture { ks }
        }
    }

    #[test]
    fn test_seal_and_open_success() {
        let f = Fixture::new(5, 3);
        let msg = make_msg("HoneyBadger BFT is robust!");
        let ct = seal(&f.ks.public_params.master_public_key, msg);

        let shares: Vec<_> = [1usize, 3, 5]
            .iter()
            .map(|&id| partial_open(&f.ks.private_shares[id - 1], &ct).unwrap())
            .collect();

        let opened = open(&f.ks.public_params, &ct, &shares).unwrap();
        assert_eq!(opened, msg);
    }

    #[test]
    fn test_open_fails_invalid_share() {
        let f = Fixture::new(5, 3);
        let msg = make_msg("This should not decrypt");
        let ct = seal(&f.ks.public_params.master_public_key, msg);

        let mut shares: Vec<_> = [1usize, 2]
            .iter()
            .map(|&id| partial_open(&f.ks.private_shares[id - 1], &ct).unwrap())
            .collect();

        // Forge a bad share for player 3
        let bad = PartialDecryptionShare {
            player_id: 3,
            value: G1::generator(),
        };
        shares.push(bad);

        assert!(open(&f.ks.public_params, &ct, &shares).is_err());
    }

    #[test]
    fn test_open_fails_insufficient_shares() {
        let f = Fixture::new(5, 3);
        let msg = make_msg("Not enough shares");
        let ct = seal(&f.ks.public_params.master_public_key, msg);

        let shares: Vec<_> = [1usize, 2]
            .iter()
            .map(|&id| partial_open(&f.ks.private_shares[id - 1], &ct).unwrap())
            .collect();

        assert!(open(&f.ks.public_params, &ct, &shares).is_err());
    }

    #[test]
    fn test_open_fails_duplicate_player() {
        let f = Fixture::new(5, 3);
        let msg = make_msg("Duplicate share");
        let ct = seal(&f.ks.public_params.master_public_key, msg);

        let s = partial_open(&f.ks.private_shares[0], &ct).unwrap();
        let shares = vec![s.clone(), s.clone(), s];

        assert!(open(&f.ks.public_params, &ct, &shares).is_err());
    }
}
